docs/e2e_testing.md

```diff
@@ -0,0 +1,398 @@
+                    schema='count:INTEGER, fee_sum:FLOAT',
+                    write_disposition=BigQueryDisposition.WRITE_TRUNCATE,
+                    additional_bq_parameters={'timePartitioning': {
+                        'type': 'DAY'
+                    }}
+                ))
+
+
+class TransactionsAggregateFn(beam.CombineFn):
+    def create_accumulator(self, *args, **kwargs):
+        return 0.0, 0
+
+    def add_input(self, mutable_accumulator, element: Transaction, *args, **kwargs):
+        fee_sum, count = mutable_accumulator
+        return fee_sum + element.fee, count + 1
+
+    def merge_accumulators(self, accumulators, *args, **kwargs):
+        fee_sums, counts = zip(*accumulators)
+        return sum(fee_sums), sum(counts)
+
+    def extract_output(self, accumulator, *args, **kwargs) -> TransactionsAggregate:
+        return TransactionsAggregate(*accumulator)
+
+
+class CalculateBitcoinAggregatesJob(bf.Job):
+    def execute(self, context: bf.JobContext):
+        p = get_dataflow_pipeline(pipeline_config)
+        aggregates_table_id = f'{dataset_config["project_id"]}:{dataset_config["dataset_name"]}.{BTC_AGGREGATES_TABLE_NAME}'
+        runtime_month = month(context.runtime_str)
+        (p
+         | 'ReadBitcoinTransactions' >> ReadBitcoinTransactions(runtime_month, external_tables_config['btc_transactions'])
+         | 'CalculateBitcoinAggregatesPTransform' >> beam.CombineGlobally(TransactionsAggregateFn())
+         | 'SaveBitcoinTransactionAggregates' >> SaveBitcoinTransactionAggregates(
+                    runtime_month, aggregates_table_id))
+
+        p.run().wait_until_finish()
+
+
+btc_aggregates_workflow = bf.Workflow(
+    workflow_id='btc_aggregates_df_bq',
+    schedule_interval='@monthly',
+    definition=[
+        CalculateBitcoinAggregatesJob()
+    ])
+```
+
+There are two facts that you should pay special attention to:
+
+* The `e2e` environment is the default one. If you import the workflow without explicitly setting the environment, it is configured using the default `e2e` environment.
+* The `e2e` configuration ensures that each execution of the workflow uses a fresh, unique BigQuery dataset (of course,
+only if you execute the workflow in a separate process).
+
+Now, take a look at the e2e test for that workflow, where we use these facts:
+
+```python
+import unittest
+from datetime import datetime, timedelta
+
+from bigflow.testing import SpawnIsolateMixin
+
+from btc_aggregates.btc_aggregates_df_bq import (
+    btc_aggregates_workflow,
+    dataset as btc_aggregates_dataset,
+    BTC_TRANSACTIONS_TABLE_NAME)
+
+DATE = slice(0, 10)
+NOW = "2020-12-01 00:00:00"
+NOW_DT = datetime.fromisoformat(NOW + "+00:00")
+NOW_MINUS_ONE_MONTH = (NOW_DT - timedelta(weeks=4)).isoformat()[DATE]
+NOW_PLUS_ONE_MONTH = (NOW_DT + timedelta(weeks=4)).isoformat()[DATE]
+
+
+class BitcoinAggregatesWorkflowTestCase(SpawnIsolateMixin, unittest.TestCase):
```

Comment:
class BitcoinAggregatesWorkflowTestCase(SpawnIsolateMixin):

Reply:
Mixin by definition shouldn't be a base class.

Reply:
Right, so maybe we could create `class EndToEndTest(SpawnIsolateMixin, unittest.TestCase)` and then use is in tests?I think it will be cleaner solution

Reply:
Do you mean adding the `EndToEndTest` class to the bigflow codebase?

Reply:
Yes

Reply:
Then add it as an issue, I don't want to go through the release cycle for such a small change.

Reply:
Ok